data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java

```diff
@@ -1022,6 +1024,30 @@ protected void execute() throws SQLException {
         }
     }
 
+    /**
+     * Gets the identity column index from the generated keys result set. In some JDBC implementations all columns are returned.
+     *
+     * @param resultSet the generated keys result set
+     * @param identity the identity persistent field
+     * @return the identity field index in the result set. If not matched then will throw {@link DataAccessException}
+     * @throws SQLException can be thrown when reading result set metadata or columns
+     */
+    private <T> int getIdentityColumnIndex(@NonNull ResultSet resultSet, RuntimePersistentProperty<?> identity) throws SQLException {
+        ResultSetMetaData resultSetMetaData = resultSet.getMetaData();
+        int columnCount = resultSetMetaData.getColumnCount();
```

Comment:
So it appears that generated keys resultset in some JDBC implementations (postgres) returns all fields, not just id and when id is not in first position then we would get an error. I tried first using `ColumnIndexResultSetReader` but it threw an error for MySql so I went with this approach.

Reply:
Can we make this dynamic so it doesn't impact performance on databases that don't return all fields? In other words if the dialect is POSTGRES use this logic, otherwise use the previous more optimized logic

Reply:
Well, essentially we are doing it with this check
```
if (columnCount == 1) {
    return 1;
}
```
if only 1 column is returned we return index 1 which is the id field like it was previously hard coded. Else, we iterate through column and match by the identity name.